### ISSUE TITLE
Dying kernel on Jupyter lab for both GPU and CPU

### ISSUE THREAD

**Issue by miguel2488** [catboost/catboost#1894]
I'm using Optuna to tune hyper parámeters of a Catboost classifier. From time to time the kernel of my notebook randomly dies before the optimization process is finished. I can't think of something else more but the parameter grid being used during some training rounds to be problem. It works well till it gets to train with a certain combination of parameters.

I have preprared a sample code snippet for both GPU and CPU:

## GPU

```
from sklearn.utils import class_weight
from catboost import Pool
import catboost as ctb
import numpy as np
import pandas as pd

# Load the data
y_train = pd.read_csv('y_train.csv').values.ravel()
y_test = pd.read_csv('y_test.csv').values.ravel()
train = pd.read_parquet('train_more_features.parquet')
test = pd.read_parquet('test_more_features.parquet')

# Class weights of the train set.
tr_weights = class_weight.compute_class_weight(class_weight = 'balanced', classes = np.unique(y_train), y = y_train)
# Weights per record on train.
tr_weight_col = np.where(y_train == 0, tr_weights[0], 
    np.where(y_train == 1, tr_weights[1],
            np.where(y_train == 2, tr_weights[2], np.nan)))
# Class weights of the test set.
ts_weights = class_weight.compute_class_weight(class_weight = 'balanced', classes = np.unique(y_test), y = y_test)
# Weights per record on test.
ts_weight_col = np.where(y_test == 0, ts_weights[0], 
    np.where(y_test == 1, ts_weights[1],
            np.where(y_test == 2, ts_weights[2], np.nan)))
# Training parameters
params = {'learning_rate': 1.9204344336384778,
          'depth': 12,
          'auto_class_weights': 'Balanced',
          'od_type': 'Iter',
          'border_count': 255,
          'feature_border_type': 'GreedyLogSum',
          'loss_function': 'MultiClass',
          'leaf_estimation_method': 'Gradient',
          'score_function': 'L2',
          'l2_leaf_reg': 27,
          'random_strength': 0.26779070318061,
          'min_child_samples': 10,
          'leaf_estimation_iterations': 3,
          'fold_permutation_block': 6,
          'fold_len_multiplier': 4,
          'boosting_type': 'Plain',
          'bootstrap_type': 'Poisson',
          'name': 'Catboost hypePar optimization',
          'verbose': 1,
          'task_type': 'GPU',
          'devices': '0',
          'thread_count': -1,
          'random_seed': 24,
          'eval_metric': 'TotalF1:average=Micro',
          'classes_count': 3,
          'num_boost_round': 500,
          'leaf_estimation_backtracking': 'No',
          'grow_policy': 'SymmetricTree'}

# Training pools
d_train = Pool(train, label = y_train, weight = tr_weight_col)
d_valid = Pool(test, label = y_test, weight = ts_weight_col)
watchlist = [d_train, d_valid]
logger.info('training Catboost:')
# Instantiate and train the model
ctb_model_opt = ctb.CatBoost(params = params)
ctb_model_opt.fit(d_train,
                  use_best_model=True,
                  eval_set=watchlist if params['task_type'] != 'GPU' else d_valid,
                  plot=False,
                  verbose_eval=False,
                  early_stopping_rounds=50
                  )
```

## CPU

```
from sklearn.utils import class_weight
from catboost import Pool
import catboost as ctb
import numpy as np
import pandas as pd

# Load the data
y_train = pd.read_csv('y_train.csv').values.ravel()
y_test = pd.read_csv('y_test.csv').values.ravel()
train = pd.read_parquet('train_more_features.parquet')
test = pd.read_parquet('test_more_features.parquet')

# Class weights of the train set.
tr_weights = class_weight.compute_class_weight(class_weight = 'balanced', classes = np.unique(y_train), y = y_train)
# Weights per record on train.
tr_weight_col = np.where(y_train == 0, tr_weights[0], 
    np.where(y_train == 1, tr_weights[1],
            np.where(y_train == 2, tr_weights[2], np.nan)))
# Class weights of the test set.
ts_weights = class_weight.compute_class_weight(class_weight = 'balanced', classes = np.unique(y_test), y = y_test)
# Weights per record on test.
ts_weight_col = np.where(y_test == 0, ts_weights[0], 
    np.where(y_test == 1, ts_weights[1],
            np.where(y_test == 2, ts_weights[2], np.nan)))
# Training parameters
params = {'learning_rate': 0.466934631791177,
          'depth': 7,
          'auto_class_weights': 'Balanced',
          'od_type': 'Iter',
          'border_count': 140,
          'feature_border_type': 'Median',
          'loss_function': 'MultiClass',
          'leaf_estimation_method': 'Newton',
          'score_function': 'Cosine',
          'l2_leaf_reg': 2,
          'random_strength': 0.001298564931328615,
          'min_child_samples': 13,
          'leaf_estimation_iterations': 6,
          'fold_permutation_block': 2,
          'fold_len_multiplier': 3,
          'boosting_type': 'Plain',
          'bootstrap_type': 'No',
          'name': 'Catboost hypePar optimization',
          'verbose': 1,
          'task_type': 'CPU',
          'devices': '0',
          'thread_count': -1,
          'random_seed': 24,
          'eval_metric': 'TotalF1:average=Micro',
          'classes_count': 3,
          'num_boost_round': 500,
          'rsm': 0.10607753982550006,
          'langevin': True,
          'leaf_estimation_backtracking': 'AnyImprovement',
          'grow_policy': 'Depthwise',
          'posterior_sampling': True}

# Training pools
d_train = Pool(train, label = y_train, weight = tr_weight_col)
d_valid = Pool(test, label = y_test, weight = ts_weight_col)
watchlist = [d_train, d_valid]
logger.info('training Catboost:')
# Instantiate and train the model
ctb_model_opt = ctb.CatBoost(params = params)
ctb_model_opt.fit(d_train,
                  use_best_model=True,
                  eval_set=watchlist if params['task_type'] != 'GPU' else d_valid,
                  plot=False,
                  verbose_eval=False,
                  early_stopping_rounds=50
                  )
```
The train data can be found [here](https://drive.google.com/file/d/1bG82Vo81SZ6uc7yxpJvWvz2ZElg6qYwK/view?usp=sharing) and the test data [here](https://drive.google.com/file/d/1_SX-RrTpqejuTCjQ6sioi_bZuiE55Pwc/view?usp=sharing).

Train labels [here](https://docs.google.com/spreadsheets/d/1encSYlmK_7Ca2RvmHAbfRyVKxIWJ215PTIvswf4GHVU/edit?usp=sharing).

Test labels [here](https://docs.google.com/spreadsheets/d/1Zo2_iW4yG_KBTuRaN_xMgGQXYvJWs413G-Ot0u-ASyo/edit?usp=sharing)

I have also tried versions 0.26, 0.26.1, 0.25.1, 0.25.0 and 0.24.4 and the problem still happens.

This problem has become really annoying, i can't finish the optimization process neither on GPU and CPU, please help.

**UPDATE**

When running on Google Colab pro with a NVIDIA P100 this is what i can see in the notebook's log:

![image](https://user-images.githubusercontent.com/42622963/138659280-e9815f05-2c43-4dda-8d95-9bcbcabe3ea0.png)

catboost version: 1.0.0
Operating System: Windows 10 Pro
GPU: NVIDIA RTX 3090
CPU: Intel(R) Core(TM) i7-5820K CPU @ 3.30GHz   3.30 GHz
RAM: 64 gb
